src/robot/output/listeners.py

```diff
@@ -15,298 +15,591 @@
+            'id': data.id,
+            'doc': result.doc,
+            'tags': list(result.tags),
+            'lineno': data.lineno,
+            'starttime': result.starttime,
+            'longname': result.full_name,
+            'source': str(data.source or ''),
+            'template': data.template or '',
+            'originalname': data.name
+        }
+        if is_end:
+            attrs.update({
+                'endtime': result.endtime,
+                'elapsedtime': result.elapsedtime,
+                'status': result.status,
+                'message': result.message,
+            })
+        return attrs
+
+    def _for_v2_attributes(self, data: 'running.For', result: 'result.For', is_end=False):
+        attrs = self._body_item_v2_attributes(data, result, is_end=is_end)
+        attrs.update({
+            'variables': list(result.assign),
+            'flavor': result.flavor or '',
+            'values': list(result.values)
+        })
+        if result.flavor == 'IN ENUMERATE':
+            attrs['start'] = result.start
+        elif result.flavor == 'IN ZIP':
+            attrs['fill'] = result.fill
+            attrs['mode'] = result.mode
+        return attrs
+
+    def _while_v2_attributes(self, data: 'running.While', result: 'result.While', is_end=False):
+        attrs = self._body_item_v2_attributes(data, result, is_end=is_end)
+        attrs.update({
+            'condition': result.condition,
+            'limit': result.limit,
+            'on_limit_message': result.on_limit_message
+            # FIXME: Add 'on_limit'
+        })
+        return attrs
+
+    def _if_v2_attributes(self, data: 'running.If', result: 'result.If', is_end=False):
+        attrs = self._body_item_v2_attributes(data, result, is_end=is_end)
+        if result.type in (BodyItem.IF, BodyItem.ELSE_IF):
+            attrs['condition'] = result.condition
+        return attrs
+
+    def _try_v2_attributes(self, data: 'running.Try', result: 'result.TryBranch', is_end=False):
+        attrs = self._body_item_v2_attributes(data, result, is_end=is_end)
+        if result.type == BodyItem.EXCEPT:
+            attrs.update({
+                'patterns': list(result.patterns),
+                'pattern_type': result.pattern_type,
+                'variable': result.assign
+            })
+        return attrs
+
+    def _return_v2_attributes(self, data: 'running.Return', result: 'result.Return', is_end=False):
+        attrs = self._body_item_v2_attributes(data, result, is_end=is_end)
+        attrs['values'] = list(result.values)
+        return result._log_name, attrs
+
+    def _body_item_v2_attributes(self, data, result, is_end=False, is_keyword_like=False):
+        attrs = {
+            'doc': result.doc,
+            'lineno': data.lineno,
+            'type': result.type,
+            'status': result.status,
+            'starttime': result.starttime,
+            'source': str(data.source or '')
+        }
+        if is_keyword_like:
```

Comment:
It would be more straightforward if keywords had a dedicated method for getting attributes. Related to that, control structures don't anymore have `doc` so with them it could be just `''`.